src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,285 @@
+        // private const byte REP_NETWORK_UNREACHABLE = 3;
+        // private const byte REP_HOST_UNREACHABLE = 4;
+        // private const byte REP_CONNECTION_REFUSED = 5;
+        // private const byte REP_TTL_EXPIRED = 6;
+        // private const byte REP_CMD_NOT_SUPPORT = 7;
+        // private const byte REP_ATYP_NOT_SUPPORT = 8;
+        private const byte CD_SUCCESS = 90;
+
+        public static ValueTask EstablishSocksTunnelAsync(Stream stream, string host, int port, Uri proxyUri, ICredentials? proxyCredentials, CancellationToken cancellationToken)
+        {
+            if (string.Equals(proxyUri.Scheme, "socks5", StringComparison.OrdinalIgnoreCase))
+            {
+                return EstablishSocks5TunnelAsync(stream, host, port, proxyUri, proxyCredentials, cancellationToken);
+            }
+            else if (string.Equals(proxyUri.Scheme, "socks4a", StringComparison.OrdinalIgnoreCase))
+            {
+                return EstablishSocks4TunnelAsync(stream, true, host, port, proxyUri, proxyCredentials, cancellationToken);
+            }
+            else if (string.Equals(proxyUri.Scheme, "socks4", StringComparison.OrdinalIgnoreCase))
+            {
+                return EstablishSocks4TunnelAsync(stream, false, host, port, proxyUri, proxyCredentials, cancellationToken);
+            }
+            else
+            {
+                Debug.Fail("Bad socks version.");
+                return default;
+            }
+        }
+
+        private static async ValueTask EstablishSocks5TunnelAsync(Stream stream, string host, int port, Uri proxyUri, ICredentials? proxyCredentials, CancellationToken cancellationToken)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(BufferSize);
+
+            try
+            {
+                // https://tools.ietf.org/html/rfc1928
+
+                // +----+----------+----------+
+                // |VER | NMETHODS | METHODS  |
+                // +----+----------+----------+
+                // | 1  |    1     | 1 to 255 |
+                // +----+----------+----------+
+                buffer[0] = ProtocolVersion5;
+                var credentials = proxyCredentials?.GetCredential(proxyUri, "");
+                if (credentials != null)
+                {
+                    buffer[1] = 1;
+                    buffer[2] = METHOD_NO_AUTH;
+                }
+                else
+                {
+                    buffer[1] = 2;
+                    buffer[2] = METHOD_NO_AUTH;
+                    buffer[3] = METHOD_USERNAME_PASSWORD;
+                }
+                await stream.WriteAsync(buffer.AsMemory(0, buffer[1] + 2), cancellationToken).ConfigureAwait(false);
+
+                // +----+--------+
+                // |VER | METHOD |
+                // +----+--------+
+                // | 1  |   1    |
+                // +----+--------+
+                await stream.ReadAsync(buffer.AsMemory(0, 2), cancellationToken).ConfigureAwait(false);
```

Comment:
`ReadAsync` isn't guaranteed to read the specified amount of bytes.

It will also return `0` on EOF, so that has to be handeled. You can use something like:

```c#
private static async ValueTask ReadAsync(Stream stream, Memory<byte> memory, bool async)
{
    while (memory.Length != 0)
    {
        int bytesRead = async
            ? await stream.ReadAsync(memory).ConfigureAwait(false)
            : stream.Read(memory.Span);

        if (bytesRead == 0)
            throw new IOException("Received premature EOF");

        memory = memory.Slice(bytesRead);
    }
}
```

Reply:
Yes. I did wrap the read-until-enough-bytes pattern when writing my own library. I'll find the pattern in BCL and follow them.